### PR TITLE
Implement xdebug.discover_client_host, fix bug with WSL2 docker-inside getting correct host.docker.internal, fixes #4803

### LINKED ISSUE #4803: WSL2 users sometimes have trouble with xdebug configuration
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Run a Diagnostic and Paste Link Here

_No response_

### Expected Behavior

Xdebug should work OOTB for WSL2 users.

### Actual Behavior

Although it works most of the time, there are some users that report that `host.docker.internal` is not properly derived, and so they end up with "connection refused", etc.

* https://github.com/ddev/ddev/issues/3772#issuecomment-1496761917 @raphaelcangucu reports working around this with `wsl hostname -I` and using that for xdebug.client_host
* https://github.com/ddev/ddev/pull/4742 - @JurriaanK reported success with using `xdebug.discover_client_host=1`

### Steps To Reproduce

_No response_

### Anything else?

xdebug.discover_client_host=1 seems like it may have general possibilities, as long as it doesn't break anything.

### PR DESCRIPTION
**Pull request by rfay**
## The Issue

* #4803 

There is always another edge case where xdebug seems not to work for people, and there was one here for xdebug with IDE on workstation and docker-ce inside WSL2.

## How This PR Solves The Issue

* Enable[xdebug.discover_client_host](https://xdebug.org/docs/all_settings#discover_client_host), which will help these situations in general
* Fix bug where UseHostDockerInternalExtraHosts was being enabled as if we were on plain linux, which meant that the wrong host.docker.internal was chosen. It should have been getting host.docker.internal from /etc/resolv.conf (for the Windows-side IP address) but instead went ahead and used `host.docker.internal:host-gateway`, resulting in behavior like where the IDE is running inside WSL2.

## Manual Testing Instructions

* Use xdebug anywhere.
* Use xdebug with IDE on Windows and docker inside WSL2

## Automated Testing Overview

There's still a need to test this exact situation, but it may be difficult because the listener has to be on Windows.

## Release/Deployment Notes

<a href="https://gitpod.io/#https://github.com/ddev/ddev/pull/4852"></a>

### CODE REVIEW

File: docs/content/users/debugging-profiling/step-debugging.md [hunk 1]
```diff
@@ -128,6 +128,16 @@ Here are basic steps to take to sort out any difficulty:
 * `ddev ssh` and try the `telnet host.docker.internal 9003` again. It should connect. If not, maybe PhpStorm is not listening, or not configured to listen on port 9003?
 * Check to make sure that Xdebug is enabled. You can use `php -i | grep -i xdebug` inside the container, or use any other technique you want that gives the output of `phpinfo()`, including Drupal’s `admin/reports/status/php`. You should see `with Xdebug v3` and `php -i | grep xdebug.mode` should give you `xdebug.mode => debug,develop => debug,develop"`.
 * Set a breakpoint in the first relevant line of your `index.php` and then visit the site in a browser. It should stop at that first line.
+* If you're using a flavor of IDE that connects directly into the web container like VS Code Language Server, you may want to use the [global configuration xdebug_ide_location](../configuration/config.md#xdebugidelocation) to explain to DDEV the situation. For example, `ddev config global --xdebug-ide-location=container`, which tells the PHP/Xdebug to connect directly to the listener inside the container.
```
Comment (mattstein) [inline docs/content/users/debugging-profiling/step-debugging.md:131]: I’d wrap the config setting in backticks and refer to it as a “setting” in an effort to use that language consistently and help subtly reinforce how you use DDEV.

```suggestion
* If you're using a flavor of IDE that connects directly into the web container like VS Code Language Server, you may want to use the [global `xdebug_ide_location` setting](../configuration/config.md#xdebugidelocation) to explain to DDEV the situation. For example, `ddev config global --xdebug-ide-location=container`, which tells the PHP/Xdebug to connect directly to the listener inside the container.
```

File: docs/content/users/debugging-profiling/step-debugging.md [hunk 2]
```diff
@@ -128,6 +128,16 @@ Here are basic steps to take to sort out any difficulty:
...
+
+### WSL2 Xdebug Troubleshooting
+
+WSL2 is a complicated environment for Xdebug, especially if you're running your IDE on the Windows side, as most people do.
+
+* With PhpStorm, consider using the "Remote Development" feature to connect to WSL. That runs an acrtual PhpStorm instance on WSL2 so the network complications are not as great.
```
Comment (mattstein) [inline docs/content/users/debugging-profiling/step-debugging.md:137]: Fix typo, rephrase to describe the affirmative.

```suggestion
* With PhpStorm, consider using the "Remote Development" feature to connect to WSL. That runs an actual PhpStorm instance on WSL2 to reduce networking complexity.
```

File: docs/content/users/debugging-profiling/step-debugging.md [hunk 3]
```diff
@@ -128,6 +128,16 @@ Here are basic steps to take to sort out any difficulty:
...
+* When using an IDE inside WSL2 (for example, when running PhpStorm or Visual Studio Code inside WSL2, or when using PhpStorm's "Remote Development" feature) you may need to tell Xdebug about this situation with `ddev config global --xdebug-ide-location=wsl2`, see [xdebug_ide_location](../configuration/config.md#xdebugidelocation).
```
Comment (mattstein) [inline docs/content/users/debugging-profiling/step-debugging.md:138]: Make this slightly easier to read, mention the _setting_ up front and in backticks, try to be more specific about what the setting is doing (which I hope I got right):

```suggestion
* When using an IDE inside WSL2—like you would running PhpStorm or Visual Studio Code inside WSL2, or using PhpStorm's "Remote Development" feature—you may need to use the [`xdebug_ide_location`](../configuration/config.md#xdebugidelocation) setting to tell Xdebug to expect the IDE under WSL2. You can do this by running `ddev config global --xdebug-ide-location=wsl2`.
```

File: docs/content/users/debugging-profiling/step-debugging.md [hunk 4]
```diff
@@ -128,6 +128,16 @@ Here are basic steps to take to sort out any difficulty:
...
+* `export DDEV_DEBUG=true && ddev start` will show you how DDEV is calculating the `host.docker.internal` IP address to be used when contacting the IDE, which may give a hint about problems you might discover in the general troubleshooting discussed above, when trying to connect to the listening IDE.
+* If you're using the IDE on the Windows side, the WSL2 file /etc/resolv.conf is supposed to be the authoritative way to figure out where the IDE is in the Windows networking scheme. That value should be the same as `host.docker.internal`, so `ddev exec ping -c 1 host.docker.internal` will show you what's actually being used. If your IDE is actually at a different address, you can tell DDEV to override the discovered value for `host.docker.internal` with `ddev config global --xdebug-ide-location=<some_ip_address>`.
```
Comment (mattstein) [inline docs/content/users/debugging-profiling/step-debugging.md:140]: Backticks, improve flow, run commands:

```suggestion
* If you're using the IDE on the Windows side, WSL2’s  `/etc/resolv.conf` file is supposed to be the authoritative way to figure out where the IDE is in the Windows networking scheme. That value should be the same as `host.docker.internal`, so running `ddev exec ping -c 1 host.docker.internal` will show you what's actually being used. If your IDE is actually at a different address, you can tell DDEV to override the discovered value for `host.docker.internal` by running `ddev config global --xdebug-ide-location=<some_ip_address>`.
```

File: docs/content/users/install/ddev-installation.md [hunk 5]
```diff
@@ -143,6 +143,7 @@ Once you’ve [installed a Docker provider](docker-installation.md), you’re re
       All Windows 10/11 editions, including Windows 10 Home support WSL2.
     * WSL2 offers a faster, smoother experience.  
       It’s vastly more performant, and you’re less likely to have obscure Windows problems.
+    * Execute DDEV commands inside WSL2 (inside Ubuntu, for example). You will never execute DDEV comamnds on the Windows side, in PowerShell or Git Bash, for example.
```
Comment (mattstein) [inline docs/content/users/install/ddev-installation.md:146]: Follow the format used with the rest of this list and avoid the typo:

```suggestion
    * Execute DDEV commands inside WSL2.   
    You’ll want to run DDEV commands inside Ubuntu, for example, and never on the Windows side in PowerShell or Git Bash.
```

**Review by mattstein — CHANGES_REQUESTED** [review]

**Comment by rfay** [comment]
Thanks so much for the quick review @mattstein !

File: docs/content/users/install/ddev-installation.md [hunk 6]
```diff
@@ -145,7 +145,7 @@ Once you’ve [installed a Docker provider](docker-installation.md), you’re re
       It’s vastly more performant, and you’re less likely to have obscure Windows problems.
 
     * Execute DDEV commands inside WSL2.   
-      You’ll want to run DDEV commands inside Ubuntu, for example, and never on the Windows side in PowerShell or Git Bash.
+      u:0You’ll want to run DDEV commands inside Ubuntu, for example, and never on the Windows side in PowerShell or Git Bash.
```
Comment (mattstein) [inline docs/content/users/install/ddev-installation.md:148]: @rfay This `u:0` doesn’t seem like it’s for humans.

Comment (mattstein) [reply docs/content/users/install/ddev-installation.md]: `z:4`! 😠

**Pull request merged**